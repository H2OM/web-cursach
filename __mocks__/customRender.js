import ClientContext from '@/lib/context/ClientContext';
import { render } from '@testing-library/react';

export default customRender = (ui, {providerProps}) => {
    return render(
      <ClientContext.Provider value={{...providerProps}}>{ui}</ClientContext.Provider>,
    )
}