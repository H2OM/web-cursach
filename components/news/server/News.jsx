import Link from 'next/link';
import Image from 'next/image';
import '../news.scss';
import Uptitle from '@/lib/basecomponents/uptitle/uptitle';
import NewsBlock from './NewsBlock';
import LazyLoad from '../client/LazyLoad';
import GET_DATA from "@/lib/GETDATA/GET_DATA";

export default async function News ({full = false, searchParams = false}) {
    const count = (searchParams.count ?? 3);
    const data = await GET_DATA({
        controller: 'news',
        action: full ? `get-news?count=${count}` : 'last-news'
    });

    return (
        <section className="News">
            <div className="container">
                <Uptitle full={full}><h2 className="title">Новости</h2></Uptitle>
                <div className="News__blocks">
                    <NewsBlock data={data}/>
                    {full ? <LazyLoad val={data.length} /> : null}
                </div>
                {
                    !full ? 
                        <div className="News__end">
                            <Link href={"/news"} className="News__end__link">Смотреть все новости</Link>
                            <Image
                                src={"/svg/vector_arrow.svg"}
                                alt="arrow"
                                className="News__end__arrow"
                                width={20}
                                height={13}
                                sizes='100vw'
                                priority={true}
                                quality={100}
                                role='arrow'
                            />
                        </div>
                    : null
                }
            </div>
        </section>
    )

}