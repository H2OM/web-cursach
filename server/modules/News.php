<?php

    namespace app\modules;

    use app\Db;

    class _News32 extends Runner {

        protected function lastNews() {

            try {

                $result = Db::getQuery("SELECT * FROM `news` ORDER BY date DESC LIMIT 3;");

                exit(json_encode($result));

            } catch (\Exception $e) {

                header("HTTP/1.0 500 Internal Server Error");

                die;
            }
        }

        protected function getNews() {

            if(isset($_GET['count']) && !preg_match('/^[0-9]*$/', $_GET['count']))  {

                header("HTTP/1.0 400 Bad Request");
                
                die;
            }

            try {

                $result = Db::getPreparedQuery("SELECT * FROM `news` ORDER BY date DESC LIMIT ?", [["VALUE"=>$_GET['count'] ?? 3, "INT"=> true]]);

                exit(json_encode($result));

            } catch (\Exception $e) {

                header("HTTP/1.0 500 Internal Server Error");
                
                die;
            }
        }
    }