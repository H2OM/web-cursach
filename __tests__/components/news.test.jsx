import customRender from "@/__mocks__/customRender";
import News from "@/components/news/server/News";
import 'intersection-observer';
import { screen } from "@testing-library/react";
import settingFetchResult from "@/__mocks__/fetchMock";
import { usePathname, useSearchParams, useRouter } from 'next/navigation';

jest.mock('next/navigation');
useRouter.mockReturnValue({
    push: jest.fn(),
}); 
usePathname.mockReturnValue("localhost");
useSearchParams.mockReturnValue({});

describe('news',()=>{

    it('full news', async ()=>{
        const titles = ['some1', 'some2', 'some3'];
        settingFetchResult([
            {
                date: "22.02.22",
                title: titles[0],
                text: "sometext",
                image: "/img/image.png"
            },
            {
                date: "22.02.52",
                title: titles[1],
                text: "sometext2",
                image: "/img/image.png"
            },
            {
                date: "22.02.24",
                title: titles[2],
                text: "sometex4t",
                image: "/img/image.png"
            },
        ]);
        const {container} = customRender(await News({full: true}), {});
        for(let title of titles) {
            expect(await screen.findByText(title)).toBeInTheDocument();
        }
        expect(container).toMatchSnapshot();
    });
    it('lite news', async ()=>{
        settingFetchResult([]);
        customRender(await News({full: false}), {});
        expect(await screen.findByText('Смотреть все новости')).toBeInTheDocument();
        expect(await screen.findByRole('arrow')).toBeInTheDocument();
    });
    it('full news exeptions', async ()=>{
        settingFetchResult([]);
        customRender(await News({full: true}), {});
        expect(await screen.findByText('Нет новостей')).toBeInTheDocument();
        expect(await screen.findByRole('anchor')).toBeInTheDocument();
        settingFetchResult(false);
        customRender(await News({full: true}), {});
        expect(await screen.findByText('Ошибка загрузки')).toBeInTheDocument();
    });
});

