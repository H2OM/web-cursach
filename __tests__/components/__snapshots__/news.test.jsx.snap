// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`news full news 1`] = `
<div>
  <section
    class="News"
  >
    <div
      class="container"
    >
      <div
        class="Uptitle__wrap"
      >
        <div
          class="Uptitle__anchorUpUptitle__anchorUp_full "
        />
        <div
          class="Uptitle Uptitle_full "
        >
          <h2
            class="title"
          >
            Новости
          </h2>
        </div>
        <div
          class="Uptitle__anchorDownUptitle__anchorDown_full "
        />
      </div>
      <div
        class="News__blocks"
      >
        <div
          class="News__blocks__block"
        >
          <div
            class="News__blocks__block__split"
          >
            <h3
              class="subtitle"
            >
              some1
            </h3>
            <div
              class="News__blocks__block__date"
            >
              22.02.22
            </div>
          </div>
          <div
            class="News__blocks__block__split"
          >
            <div
              class="News__blocks__block__desc"
            >
              sometext
            </div>
            <img
              alt="image"
              class="News__blocks__block__image"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="0"
              sizes="100vw"
              src="/_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=3840&q=100"
              srcset="/_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=640&q=100 640w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=750&q=100 750w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=828&q=100 828w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1080&q=100 1080w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1200&q=100 1200w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1920&q=100 1920w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=2048&q=100 2048w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=3840&q=100 3840w"
              style="color: transparent;"
              width="0"
            />
          </div>
        </div>
        <div
          class="News__blocks__block"
        >
          <div
            class="News__blocks__block__split"
          >
            <h3
              class="subtitle"
            >
              some2
            </h3>
            <div
              class="News__blocks__block__date"
            >
              22.02.52
            </div>
          </div>
          <div
            class="News__blocks__block__split"
          >
            <div
              class="News__blocks__block__desc"
            >
              sometext2
            </div>
            <img
              alt="image"
              class="News__blocks__block__image"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="0"
              sizes="100vw"
              src="/_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=3840&q=100"
              srcset="/_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=640&q=100 640w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=750&q=100 750w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=828&q=100 828w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1080&q=100 1080w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1200&q=100 1200w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1920&q=100 1920w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=2048&q=100 2048w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=3840&q=100 3840w"
              style="color: transparent;"
              width="0"
            />
          </div>
        </div>
        <div
          class="News__blocks__block"
        >
          <div
            class="News__blocks__block__split"
          >
            <h3
              class="subtitle"
            >
              some3
            </h3>
            <div
              class="News__blocks__block__date"
            >
              22.02.24
            </div>
          </div>
          <div
            class="News__blocks__block__split"
          >
            <div
              class="News__blocks__block__desc"
            >
              sometex4t
            </div>
            <img
              alt="image"
              class="News__blocks__block__image"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="0"
              sizes="100vw"
              src="/_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=3840&q=100"
              srcset="/_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=640&q=100 640w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=750&q=100 750w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=828&q=100 828w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1080&q=100 1080w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1200&q=100 1200w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=1920&q=100 1920w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=2048&q=100 2048w, /_next/image?url=%2Fimg%2F%2Fimg%2Fimage.png&w=3840&q=100 3840w"
              style="color: transparent;"
              width="0"
            />
          </div>
        </div>
        <div
          class="News__blocks__loadAnchor"
          role="anchor"
        />
      </div>
    </div>
  </section>
</div>
`;
